[19.0][MIG] hr_timesheet_time_control: migrated from project/project_timesheet_time_control

CRogos opened the pull request:
Migration and changes according to roadmap.

sergio-teruel committed:
project_timesheet_time_control: Module rename and fixes

pedrobaeza committed:
project_timesheet_time_control

* Decorate methods correctly + README imp
* Add tests
* Add button on tasks
* Better README
* open/close behaviour

luismontalba committed:
[MIG] project_timesheet_time_control: Migrated to 10.0

fuentes73 committed:
[FIX] project_timesheet_time_control: Condition always gets evaluated as False (#321)

ernestotejeda committed:
[MIG] project_timesheet_time_control: Migration to 11.0

Update README
Adapt tests
Redefine the new kanban and form view of account.analytic.line
to add 'Close/Open task' button and 'Stop' button

pedrobaeza committed:
[FIX] project_timesheet_time_control: Don't reset task

We need to override standard onchange, as it resets the previously
introduced task.

ernestotejeda committed:
[MIG] project_timesheet_time_control: Migration to 12.0

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (8 of 8 strings)

Translation: project-12.0/project-12.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-12-0/project-12-0-project_timesheet_time_control/pt_BR/

yajo committed:
[IMP] project_timesheet_time_control: ❤️ 4 UX

This is a huge improvement in this module's usability, although basic functionality stays the same. Let me summmarize:

- In `account.analytic.line` views, the *Stop* button is replaced by colorful *Resume/Stop* buttons.
- The new *Resume* button will display a wizard that will duplicate all meaningful fields from the previous `account.analytic.line` record and auto-stop the previous running timer, if any.
- If the previous running timer is very old, some warnings will be displayed.
- In `project.project` and `project.task` views, I added the same buttons (well... it's *Start* instead of *Resume*, because you could not have worked before in that project or task, but it behaves the same).
- If using it from a `project.task`, *Start* will duplicate meaningful fields from your latest line in that task.
- If using it from a `project.project`, *Start* will duplicate meaningful fields your latest line in that project that had no task associated.
- All kanban, kanban mobile, list and form views are covered.
- Access the new wizard directly from a menu item if you prefer.
- I removed, BTW, some Bootstrap 3 and bad replacements in some views.

yajo committed:
[IMP] project_timesheet_time_control: Tell the user no running line was found

Imagine this scenario:

1. In tab 1 of the browser, you have opened task 1.
2. In tab 2 of the browser, you have opened task 2.
3. You go to tab 1 and start a timer.
4. Work, work, work...
5. You go to tab 2 and start a timer, stopping that of task 1.
6. Work, work, work...
7. You go to tab 1 and see that timer as running (it is not, but you didn't refresh). You hit stop.

Before this commit, it just seemed like the timer was actually stopped. What did happen behind the scenes is that your view was refreshed, but no timer was touched fortunately.

After this commit, you get a message telling you that there's no timer to stop and that your browser is most likely out of sync. This mimics the behavior previously found when doing the same, but directly in the AAL. Now it's present in projects and tasks too.

yajo committed:
[IMP] project_timesheet_time_control: Refactor with mixin

Models related to timesheet time controls now inherit from a mixin that adds most needed logic automatically.

sysadminmatmoz committed:
Translated using Weblate (Slovenian)

Currently translated at 32.3% (20 of 62 strings)

Translation: project-12.0/project-12.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-12-0/project-12-0-project_timesheet_time_control/sl/

marylla committed:
Translated using Weblate (German)

Currently translated at 7.9% (5 of 63 strings)

Translation: project-12.0/project-12.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-12-0/project-12-0-project_timesheet_time_control/de/

pedrobaeza committed:
[IMP] : black, isort, prettier

pedrobaeza committed:
[MIG] project_timesheet_time_control: Migration to 13.0

- Standard procedure
- README regeneration
- Start new timer little adjustment
- Avoid new check for shared m2m tables

davejames committed:
[FIX] project_timesheet_time_control: User timezone + comp. with hr_timesheet_activity_begin_end + field for end time

- take users timezone into account when setting date field
- Prior to this commit, trying to resume an activity which had a start
  time or end time set would cause the newly created timesheet to throw
  a validation error because time_stop - time_start must equal duration.
  This commit fixes that issues by adding those fields to the ignore list
- Adds a field to the timesheet to store the end time. It is not
  required to enter the end time, if it is not entered it is calculated
  automatically by comparing the start time and the duration, if only a
  start time and end time are entered, then the duration is calculated
  accordingly. The end time field is set to optional in the tree view.

eLBati committed:
FIX project_timesheet_time_control when not using employees

ygol committed:
Added translation using Weblate (Hebrew)

ygol committed:
Translated using Weblate (Hebrew)

Currently translated at 100.0% (67 of 67 strings)

Translation: project-13.0/project-13.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-13-0/project-13-0-project_timesheet_time_control/he_IL/

Commit:
Translated using Weblate (German)

Currently translated at 16.4% (11 of 67 strings)

Translation: project-13.0/project-13.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-13-0/project-13-0-project_timesheet_time_control/de/

bosd committed:
Translated using Weblate (Dutch)

Currently translated at 67.1% (45 of 67 strings)

Translation: project-13.0/project-13.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-13-0/project-13-0-project_timesheet_time_control/nl/

Commit:
[IMP] project_timesheet_time_control: black, isort, prettier

Commit:
[MIG] project_timesheet_time_control: Migration to 14.0

ibuioli committed:
Translated using Weblate (Spanish (Argentina))

Currently translated at 100.0% (67 of 67 strings)

Translation: project-14.0/project-14.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-14-0/project-14-0-project_timesheet_time_control/es_AR/

petrus-v committed:
[project_timesheet_time_control] 14.0 fix warning

bosd committed:
Translated using Weblate (Dutch)

Currently translated at 88.0% (59 of 67 strings)

Translation: project-14.0/project-14.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-14-0/project-14-0-project_timesheet_time_control/nl/

remi-filament committed:
Translated using Weblate (French)

Currently translated at 95.5% (64 of 67 strings)

Translation: project-14.0/project-14.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-14-0/project-14-0-project_timesheet_time_control/fr/

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 14.9% (10 of 67 strings)

Translation: project-14.0/project-14.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-14-0/project-14-0-project_timesheet_time_control/it/

manuelcalerosolis committed:
[MIG] project_timesheet_time_control: Migration to 15.0

ibuioli committed:
Translated using Weblate (Spanish (Argentina))

Currently translated at 100.0% (76 of 76 strings)

Translation: project-14.0/project-14.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-14-0/project-14-0-project_timesheet_time_control/es_AR/

victoralmau committed:
[DCK] Remove development_status to project_timesheet_time_control.

jarmokortetjarvi committed:
Translated using Weblate (Finnish)

Currently translated at 96.0% (73 of 76 strings)

Translation: project-14.0/project-14.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-14-0/project-14-0-project_timesheet_time_control/fi/

pedrobaeza committed:
[FIX] project_timesheet_time_control: Avoid texts on buttons

Since v14, `string` attribute of <button> elements are directly shown in
the views as text instead of being a tooltip, and thus, it takes a lot
of space in screen.

And it was not the intended visualization when the modules was
conceived, just a side effect.

Thus, we restore what we expect, moving `string` to `title` for having
the tooltips again.

TT40169

CarlosRoca13 committed:
[FIX] project_timesheet_time_control: Avoid incompatibilities
Some modules that add new fields dynamically, as analytic_tag_dimension, will cause an error when trying to read the fields to get them from cache. With this change we avoid that read and we make a copy of the analytic line if the wizard is opened from its and we create a new analytic line if don't.

ibuioli committed:
Translated using Weblate (Spanish (Argentina))

Currently translated at 100.0% (49 of 49 strings)

Translation: project-15.0/project-15.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-15-0/project-15-0-project_timesheet_time_control/es_AR/

macagua committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (49 of 49 strings)

Translation: project-15.0/project-15.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-15-0/project-15-0-project_timesheet_time_control/es/

ignaciord committed:
[15.0][FIX][project_timesheet_time_control]

Error: AttributeError: type object 'expression' has no attribute AND

pedrobaeza committed:
[OU-ADD] project_timesheet_time_control: Reset view inheritance

That way, we don't have an error coming from the previous version

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (49 of 49 strings)

Translation: project-15.0/project-15.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-15-0/project-15-0-project_timesheet_time_control/it/

victoralmau committed:
[IMP] project_timesheet_time_control: Add base.group_multi_company group to company_id field from wizard

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (49 of 49 strings)

Translation: project-15.0/project-15.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-15-0/project-15-0-project_timesheet_time_control/it/

aktiv-heli-kantawala committed:
[MIG] project_timesheet_time_control: Migration to 16.0

Co-Authored-By: Stefan Ungureanu <stefan.ungureanu@tecnativa.com>
Co-Authored-By: Pedro M. Baeza <pedro.baeza@tecnativa.com>

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (48 of 48 strings)

Translation: project-16.0/project-16.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-16-0/project-16-0-project_timesheet_time_control/it/

moctarjallo committed:
[16.0][FIX]project_timesheet_time_control: Fix warning on method invalidate_cache

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 87.5% (42 of 48 strings)

Translation: project-16.0/project-16.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-16-0/project-16-0-project_timesheet_time_control/pt_BR/

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (48 of 48 strings)

Translation: project-16.0/project-16.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-16-0/project-16-0-project_timesheet_time_control/pt_BR/

pere-aquarian committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (48 of 48 strings)

Translation: project-16.0/project-16.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-16-0/project-16-0-project_timesheet_time_control/ca/

jakobkrabbe committed:
Translated using Weblate (Swedish)

Currently translated at 100.0% (48 of 48 strings)

Translation: project-16.0/project-16.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-16-0/project-16-0-project_timesheet_time_control/sv/

Commit:
[IMP] project_timesheet_time_control: pre-commit auto fixes

Commit:
[MIG] project_timesheet_time_control: Migration to 17.0

WildfootW committed:
Translated using Weblate (Chinese (Traditional) (zh_TW))

Currently translated at 4.2% (2 of 47 strings)

Translation: project-17.0/project-17.0-project_timesheet_time_control
Translate-URL: https://translation.odoo-community.org/projects/project-17-0/project-17-0-project_timesheet_time_control/zh_TW/

victoralmau committed:
[FIX] project_timesheet_time_control: Avoid access error with ir.model

Only "advanced" users have read access to ir.model, for this reason, we need to apply .sudo()

jdebetaz committed:
[IMP] project_timesheet_time_control: pre-commit auto fixes

jdebetaz committed:
[MIG] project_timesheet_time_control: Migration to 18.0

[IMP] project_timesheet_time_control: pre-commit execution

[MIG] project_timesheet_time_control: Migration to 18.0

[MIG] project_timesheet_time_control: Migration to 18.0

[MIG] project_timesheet_time_control: Migration to 18.0

[MIG] project_timesheet_time_control: Migration to 18.0

david-banon-tecnativa committed:
[FIX] project_timesheet_time_control: hide show_time_control
In odoo 17+ tree/list view now uses column_invisible instead of invisible to hide the entire column.

TT55789

leblanc-simon committed:
[18.0][FIX] project_timesheet_time_control : invisible to column_invisible for tree view

pilarvargas-tecnativa committed:
[FIX] project_timesheet_time_control: Correctly hide show_time_control in task form

victoralmau committed:
[IMP] project_timesheet_time_control: Test improvements

Changes done:
- Use BaseCommon
- Define TestProjectTimesheetTimeControlBase to be able to inherit it in other modules (crm_timesheet for example)
- Don't use admin user and create a user for the tests

TT56039

victoralmau committed:
[IMP] project_timesheet_time_control: Change maintainer

CRogos committed:
[IMP] project_timesheet_time_control: sync unit_amount with date_time and date_time_end on form view

CRogos committed:
[IMP] project_timesheet_time_control: enhance date handling

CRogos committed:
[IMP] hr_timesheet_time_control: pre-commit auto fixes

CRogos committed:
[MIG] hr_timesheet_time_control: Migration to 19.0

christian-ramos-tecnativa reviewed (approved):
LGTM

christian-ramos-tecnativa commented:
This migration contains the changes in 

- [ ] https://github.com/OCA/project/pull/1588

bosd reviewed (approved):
Funcional test :+1

CRogos commented:
Updated code from PR: https://github.com/OCA/project/pull/1588

I also would like to add this PR https://github.com/OCA/project/pull/1609, but than I get this depreciation warning. 
```
2025-11-26 14:35:29,241 131608 WARNING db18_test_hr_timesheet_time_control odoo.domains: The domain condition ('date_time', '<', SQL("%s - account_analytic_line.unit_amount * (select 1 / factor * %s from uom_uom where id = account_analytic_line.product_uom_id) * interval '1 hour'", datetime.datetime(2023, 1, 10, 11, 0, 1), 1.0)) should use the 'any' or 'not any' operator. 
```

CRogos committed:
[IMP] project_timesheet_time_control: improve date_time handling in create and write methods

NICO-SOLUTIONS commented:
@CRogos 
hey. 
What do you think about including a time rounding feature with minimal duration and round up like in enterprise?
I just saw, that this module is an altenative for hr_timesheet_begin_end and i am checking, if it makes sense to switch to the new structure.

CRogos commented:
@NICO-SOLUTIONS I also changed the module with version 18 and like it alot.

We are using the enterprise version and also the rounding method, therefore we do not have a need. I think @pedrobaeza is not a fan to add this to this module, either. 

But maybe we can add it to [hr_timesheet_calendar](https://github.com/OCA/timesheet/pull/842) . There I've already added an alternate start time behavior ("timenow()" vs "after previous entry"). If implemented and configured correctly, two rounding (oca + enterprise) should not be a problem. If you don't want the calendar dependency a new module is also an option.

PS: @pedrobaeza can you merge?

CRogos commented:
@victoralmau can you merge?

ivs-cetmix commented:
Hey @CRogos thank you for your contribution!
/ocabot merge nobump
